mithril-common/src/resource_pool.rs

```diff
@@ -0,0 +1,316 @@
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+
+        let mut resources_items = vec![];
+        for _ in 0..pool_size {
+            let resource_item = pool.acquire_resource(Duration::from_millis(1000)).unwrap();
+            resources_items.push(resource_item);
+        }
+        let resources_result = resources_items
+            .iter_mut()
+            .map(|resource_item| resource_item.resource().to_owned())
+            .collect::<Vec<_>>();
+
+        assert_eq!(resources_expected, resources_result);
+        assert_eq!(pool.count().unwrap(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_resource_pool_acquire_locks_until_timeout_when_no_resource_available() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+
+        let mut resources_items = vec![];
+        for _ in 0..pool_size {
+            let resource_item = pool.acquire_resource(Duration::from_millis(1000)).unwrap();
+            resources_items.push(resource_item);
+        }
+
+        assert!(pool.acquire_resource(Duration::from_millis(1000)).is_err());
+    }
+
+    #[tokio::test]
+    async fn test_resource_pool_drains_successfully() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+        assert_eq!(pool.count().unwrap(), pool_size);
+
+        pool.drain();
+
+        assert_eq!(pool.count().unwrap(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_resource_pool_returns_fresh_resource() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+        assert_eq!(pool.count().unwrap(), pool_size);
+
+        let mut resource_item = pool.acquire_resource(Duration::from_millis(1000)).unwrap();
+        assert_eq!(pool.count().unwrap(), pool_size - 1);
+        pool.return_resource(resource_item.into_inner(), pool.discriminant().unwrap())
+            .unwrap();
+
+        assert_eq!(pool.count().unwrap(), pool_size);
+    }
+
+    #[tokio::test]
+    async fn test_resource_pool_returns_resource_automatically() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+        assert_eq!(pool.count().unwrap(), pool_size);
+
+        {
+            let _resource_item = pool.acquire_resource(Duration::from_millis(1000)).unwrap();
+            assert_eq!(pool.count().unwrap(), pool_size - 1);
+        }
```

Comment:
It's not easy to understand that the resource returns to the pool because it goes out of scope.
Maybe a comment could help for those who are not sufficiently familiar with Rust.

Reply:
Done